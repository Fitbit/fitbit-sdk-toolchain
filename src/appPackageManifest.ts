import path from 'path';
import { Transform, TransformCallback } from 'stream';

import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';
import { pipe } from 'fp-ts/lib/pipeable';
import { fold } from 'fp-ts/lib/Either';
import lodash from 'lodash';
import PluginError from 'plugin-error';
import Vinyl from 'vinyl';

import { normalizeToPOSIX } from './pathUtils';
import ProjectConfiguration from './ProjectConfiguration';
import { apiVersions } from './sdkVersion';

const manifestPath = 'manifest.json';
const PLUGIN_NAME = 'appPackageManifest';

interface Components {
  watch?: {
    [platform: string]: {
      filename: string;
      platform: string[];
    };
  };
  companion?: {
    filename: string;
  };
}

// tslint:disable-next-line:variable-name
const ComponentBundleTag = t.taggedUnion('type', [
  t.intersection([
    t.interface({
      type: t.literal('device'),
      family: t.string,
      platform: t.array(t.string),
    }),
    t.partial({
      isNative: t.literal(true),
    }),
  ]),
  t.type({
    type: t.literal('companion'),
  }),
]);
type ComponentBundleTag = t.TypeOf<typeof ComponentBundleTag>;

function getBundleInfo(file: Vinyl): ComponentBundleTag {
  return pipe(
    ComponentBundleTag.decode(file.componentBundle),
    fold(
      (errors) => {
        throw new PluginError(
          PLUGIN_NAME,
          `Unknown bundle component tag: ${failure(errors).join('\n')}`,
          { fileName: file.relative },
        );
      },
      (tag) => tag,
    ),
  );
}

class AppPackageManifestTransform extends Transform {
  sourceMaps = {};
  components: Components = {};
  hasJS = false;
  hasNative = false;

  constructor(
    public projectConfig: ProjectConfiguration,
    public buildID: string,
  ) {
    super({ objectMode: true });
  }

  private transformComponentBundle(file: Vinyl) {
    const bundleInfo = getBundleInfo(file);

    function throwDuplicateComponent(existingPath: string) {
      const componentType =
        bundleInfo.type === 'device'
          ? `${bundleInfo.type}/${bundleInfo.family}`
          : bundleInfo.type;
      throw new PluginError(
        PLUGIN_NAME,
        `Duplicate ${componentType} component bundles: ${file.relative} / ${existingPath}`,
      );
    }

    if (bundleInfo.type === 'device') {
      if (bundleInfo.isNative) this.hasNative = true;
      else this.hasJS = true;

      if (this.hasJS && this.hasNative) {
        throw new PluginError(
          PLUGIN_NAME,
          'Cannot bundle mixed native/JS device components',
          { fileName: file.relative },
        );
      }

      if (!this.components.watch) this.components.watch = {};

      if (this.components.watch[bundleInfo.family]) {
        throwDuplicateComponent(
          this.components.watch[bundleInfo.family].filename,
        );
      }

      this.components.watch[bundleInfo.family] = {
        platform: bundleInfo.platform,
        filename: file.relative,
      };
    } else {
      if (this.components[bundleInfo.type] !== undefined) {
        throwDuplicateComponent(this.components[bundleInfo.type]!.filename);
      }

      this.components[bundleInfo.type] = { filename: file.relative };
    }
  }

  // tslint:disable-next-line:function-name
  _transform(file: Vinyl, _: unknown, next: TransformCallback) {
    if (file.componentMapKey) {
      lodash.merge(
        this.sourceMaps,
        lodash.set({}, file.componentMapKey, normalizeToPOSIX(file.relative)),
      );
    }

    if (file.componentBundle) {
      try {
        this.transformComponentBundle(file);
      } catch (ex) {
        return next(ex);
      }
    }

    return next(undefined, file);
  }

  // tslint:disable-next-line:function-name
  _flush(callback: TransformCallback) {
    const setSDKVersion =
      (this.components.watch && this.hasJS) || this.components.companion;
    const { deviceApi, companionApi } = apiVersions(this.projectConfig);
    const manifestJSON = JSON.stringify(
      {
        buildId: this.buildID,
        components: this.components,
        sourceMaps: this.sourceMaps,
        manifestVersion: 6,
        ...(setSDKVersion && {
          sdkVersion: {
            ...(this.components.watch && this.hasJS && { deviceApi }),
            ...(this.components.companion && { companionApi }),
          },
        }),
        requestedPermissions: this.projectConfig.requestedPermissions,
        appId: this.projectConfig.appUUID,
      },
      undefined,
      2,
    );
    this.push(
      new Vinyl({
        contents: Buffer.from(manifestJSON, 'utf8'),
        path: path.resolve(process.cwd(), manifestPath),
      }),
    );
    callback();
  }
}

export default function appPackageManifest({
  projectConfig,
  buildId,
}: {
  projectConfig: ProjectConfiguration;
  buildId: string;
}) {
  return new AppPackageManifestTransform(projectConfig, buildId);
}
