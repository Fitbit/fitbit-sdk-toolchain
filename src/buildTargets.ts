import lodash from 'lodash';
import semver from 'semver';

import sdkVersion from './sdkVersion';
import { PolyfillMap } from './plugins/polyfill';

export interface BuildTargetDescriptor {
  displayName: string;
  platform: string[];
  resourceFilterTag: string;
  polyfills?: PolyfillMap;
  maxDeviceBundleSize?: number; // in bytes
  minSDKVersion?: string;
}

const baseBuildTargets: { [platform: string]: BuildTargetDescriptor } = {
  higgs: {
    displayName: 'Fitbit Ionic',
    platform: ['30.1.2+'],
    resourceFilterTag: '348x250',
  },
  meson: {
    displayName: 'Fitbit Versa',
    platform: ['32.4.18+'],
    resourceFilterTag: '300x300',
  },
  gemini: {
    displayName: 'Fitbit Versa Lite',
    platform: ['33.1.30+'],
    resourceFilterTag: '300x300',
    minSDKVersion: '3.1.0',
    maxDeviceBundleSize: 3145728,
    polyfills: {
      barometer: 'export var Barometer = undefined; export default Barometer;',
      gyroscope: 'export var Gyroscope = undefined; export default Gyroscope;',
      orientation:
        'export var OrientationSensor = undefined; export default OrientationSensor;',
      'user-activity': `
        export { today, goals, default } from 'user-activity';
        import { today } from 'user-activity';
        Object.defineProperty(today.local, 'elevationGain', {});
      `,
    },
  },
  mira: {
    displayName: 'Fitbit Versa 2',
    platform: ['68.9.12+'],
    minSDKVersion: '4.0.0',
    resourceFilterTag: '300x300',
  },
};

let extraBuildTargets: typeof baseBuildTargets | undefined;
try {
  extraBuildTargets = require('@fitbit/sdk-build-targets').default;
} catch {}

export function generateBuildTargets() {
  return lodash.pickBy(
    {
      ...baseBuildTargets,
      ...extraBuildTargets,
    },
    ({ minSDKVersion }) =>
      minSDKVersion === undefined ||
      semver.gte(sdkVersion().format(), minSDKVersion),
  );
}

const buildTargets = generateBuildTargets();
export default buildTargets;
