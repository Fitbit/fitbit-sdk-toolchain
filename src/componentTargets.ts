interface ComponentTarget {
  inputs: string[];
  outputDir?: string;
  notFoundIsFatal: boolean;
  translationsGlob: string;
}

export enum ComponentType {
  DEVICE = 'device',
  COMPANION = 'companion',
  SETTINGS = 'settings',
}

const componentTarget: { [component in ComponentType]: ComponentTarget } = {
  [ComponentType.DEVICE]: {
    inputs: ['app/index.ts', 'app/index.js'],
    outputDir: 'app',
    notFoundIsFatal: true,
    translationsGlob: 'app/i18n/*.po',
  },
  [ComponentType.COMPANION]: {
    inputs: ['companion/index.ts', 'companion/index.js'],
    notFoundIsFatal: false,
    translationsGlob: 'companion/i18n/*.po',
  },
  [ComponentType.SETTINGS]: {
    inputs: ['tsx', 'ts', 'jsx', 'js'].map((ext) => `settings/index.${ext}`),
    notFoundIsFatal: false,
    translationsGlob: 'settings/i18n/*.po',
  },
};

export default componentTarget;
