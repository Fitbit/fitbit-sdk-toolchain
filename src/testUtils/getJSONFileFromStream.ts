import { Readable } from 'stream';

import getFileFromStream from './getFileFromStream';
import getVinylContents from './getVinylContents';

export default function getJSONFileFromStream(
  stream: Readable,
  filename?: string,
) {
  return getFileFromStream(stream, filename)
    .then(getVinylContents)
    .then(JSON.parse);
}
