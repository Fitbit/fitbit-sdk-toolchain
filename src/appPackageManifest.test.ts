import { Readable } from 'stream';

import Vinyl from 'vinyl';
import appPackageManifest from './appPackageManifest';
import buildTargets from './buildTargets';
import { ComponentType } from './componentTargets';
import ProjectConfiguration, {
  AppType,
  ClockProjectConfiguration,
} from './ProjectConfiguration';

import getJSONFileFromStream from './testUtils/getJSONFileFromStream';

jest.mock('./packageVersion.const');

const buildId = '0x0f75775f470c1585';
const makeProjectConfig = (): ClockProjectConfiguration => ({
  appUUID: 'b4ae822e-eca9-4fcb-8747-217f2a1f53a1',
  appType: AppType.CLOCKFACE,
  appDisplayName: 'My App',
  i18n: {
    en: { name: 'My App' },
    fr: { name: 'Mon application' },
  },
  buildTargets: ['higgs'],
  requestedPermissions: [],
  defaultLanguage: 'en-US',
});

function makeReadStream() {
  const stream = new Readable({ objectMode: true });
  stream._read = () => {};
  return stream;
}

function expectPackageManifest(
  stream: Readable,
  projectConfig: ProjectConfiguration,
) {
  return expect(
    getJSONFileFromStream(
      stream.pipe(
        appPackageManifest({
          buildId,
          projectConfig,
        }),
      ),
      'manifest.json',
    ),
  );
}

function expectValidPackageManifest(options?: {
  hasCompanion?: boolean;
  projectConfig?: ProjectConfiguration;
  nativeApp?: boolean;
}) {
  const { hasCompanion, projectConfig, nativeApp } = {
    projectConfig: makeProjectConfig(),
    hasCompanion: false,
    nativeApp: false,
    ...options,
  };
  const stream = makeReadStream();
  for (const platform of projectConfig.buildTargets) {
    if (!nativeApp) {
      stream.push(
        new Vinyl({
          path: `sourceMaps/${ComponentType.DEVICE}/${platform}/index.js.json`,
          contents: Buffer.alloc(0),
          componentMapKey: [ComponentType.DEVICE, platform],
        }),
      );
    }
    stream.push(
      new Vinyl({
        path: `${ComponentType.DEVICE}-${platform}.zip`,
        contents: Buffer.alloc(0),
        componentBundle: {
          type: 'device',
          family: platform,
          platform: buildTargets[platform].platform,
          ...(nativeApp && { isNative: true }),
        },
      }),
    );
  }
  if (hasCompanion) {
    for (const componentType of [
      ComponentType.COMPANION,
      ComponentType.SETTINGS,
    ]) {
      stream.push(
        new Vinyl({
          path: `sourceMaps/${componentType}/${componentType}.js.json`,
          contents: Buffer.alloc(0),
          componentMapKey: [componentType],
        }),
      );
    }
    stream.push(
      new Vinyl({
        path: `${ComponentType.COMPANION}.zip`,
        contents: Buffer.alloc(0),
        componentBundle: {
          type: 'companion',
        },
      }),
    );
  }
  stream.push(null);
  return expectPackageManifest(stream, projectConfig).resolves;
}

it('builds a package manifest', () =>
  expectValidPackageManifest().toMatchSnapshot());

it('builds a package manifest with a companion', () =>
  expectValidPackageManifest({ hasCompanion: true }).toMatchSnapshot());

it('builds a package manifest with multiple device components', () =>
  expectValidPackageManifest({
    projectConfig: {
      ...makeProjectConfig(),
      buildTargets: ['higgs', 'meson'],
    },
  }).toMatchSnapshot());

it('emits an error if both JS and native device components are present', () => {
  const projectConfig = makeProjectConfig();
  const stream = makeReadStream();
  stream.push(
    new Vinyl({
      componentBundle: {
        type: 'device',
        family: 'foo',
        platform: ['1.1.1+'],
      },
      path: 'bundle.zip',
      contents: Buffer.alloc(0),
    }),
  );
  stream.push(
    new Vinyl({
      componentBundle: {
        type: 'device',
        family: 'bar',
        platform: ['1.1.1+'],
        isNative: true,
      },
      path: 'bundle.bin',
      contents: Buffer.alloc(0),
    }),
  );
  stream.push(null);

  return expectPackageManifest(
    stream,
    projectConfig,
  ).rejects.toThrowErrorMatchingSnapshot();
});

it('emits an error if multiple bundles are present for the same device family', () => {
  const projectConfig = makeProjectConfig();
  const stream = makeReadStream();
  for (let i = 0; i <= 3; i += 1) {
    stream.push(
      new Vinyl({
        componentBundle: {
          type: 'device',
          family: 'foo',
          platform: ['1.1.1+'],
        },
        path: `bundle${i}.zip`,
        contents: Buffer.alloc(0),
      }),
    );
  }
  stream.push(null);

  return expectPackageManifest(
    stream,
    projectConfig,
  ).rejects.toThrowErrorMatchingSnapshot();
});

it('builds a package manifest with a native device component', () =>
  expectValidPackageManifest({ nativeApp: true }).toMatchSnapshot());

it('builds a package manifest with a native device component and companion', () =>
  expectValidPackageManifest({
    nativeApp: true,
    hasCompanion: true,
  }).toMatchSnapshot());

it.each([
  ['has an invalid type field', { type: '__invalid__' }],
  ['has a device type but missing platform', { type: 'device', family: 'foo' }],
  [
    'has a device type but missing family',
    { type: 'device', platform: ['1.1.1+'] },
  ],
  [
    'has a device type but invalid platform',
    { type: 'device', family: 'foo', platform: '1.1.1+' },
  ],
])('emits an error if a component bundle tag %s', (_, componentBundle) => {
  const projectConfig = makeProjectConfig();
  const stream = makeReadStream();
  stream.push(
    new Vinyl({
      componentBundle,
      path: 'bundle.zip',
      contents: Buffer.alloc(0),
    }),
  );

  return expectPackageManifest(
    stream,
    projectConfig,
  ).rejects.toThrowErrorMatchingSnapshot();
});
