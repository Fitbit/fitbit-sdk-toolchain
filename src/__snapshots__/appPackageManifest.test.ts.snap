// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a package manifest 1`] = `
Object {
  "appId": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
  "buildId": "0x0f75775f470c1585",
  "components": Object {
    "watch": Object {
      "higgs": Object {
        "filename": "device-higgs.zip",
        "platform": Array [
          "30.1.2+",
        ],
      },
    },
  },
  "manifestVersion": 6,
  "requestedPermissions": Array [],
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": Object {
    "device": Object {
      "higgs": "sourceMaps/device/higgs/index.js.json",
    },
  },
}
`;

exports[`builds a package manifest with a companion 1`] = `
Object {
  "appId": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
  "buildId": "0x0f75775f470c1585",
  "components": Object {
    "companion": Object {
      "filename": "companion.zip",
    },
    "watch": Object {
      "higgs": Object {
        "filename": "device-higgs.zip",
        "platform": Array [
          "30.1.2+",
        ],
      },
    },
  },
  "manifestVersion": 6,
  "requestedPermissions": Array [],
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": Object {
    "companion": "sourceMaps/companion/companion.js.json",
    "device": Object {
      "higgs": "sourceMaps/device/higgs/index.js.json",
    },
    "settings": "sourceMaps/settings/settings.js.json",
  },
}
`;

exports[`builds a package manifest with a native device component 1`] = `
Object {
  "appId": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
  "buildId": "0x0f75775f470c1585",
  "components": Object {
    "watch": Object {
      "higgs": Object {
        "filename": "device-higgs.zip",
        "platform": Array [
          "30.1.2+",
        ],
      },
    },
  },
  "manifestVersion": 6,
  "requestedPermissions": Array [],
  "sourceMaps": Object {},
}
`;

exports[`builds a package manifest with a native device component and companion 1`] = `
Object {
  "appId": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
  "buildId": "0x0f75775f470c1585",
  "components": Object {
    "companion": Object {
      "filename": "companion.zip",
    },
    "watch": Object {
      "higgs": Object {
        "filename": "device-higgs.zip",
        "platform": Array [
          "30.1.2+",
        ],
      },
    },
  },
  "manifestVersion": 6,
  "requestedPermissions": Array [],
  "sdkVersion": Object {
    "companionApi": "1.0.0",
  },
  "sourceMaps": Object {
    "companion": "sourceMaps/companion/companion.js.json",
    "settings": "sourceMaps/settings/settings.js.json",
  },
}
`;

exports[`builds a package manifest with multiple device components 1`] = `
Object {
  "appId": "b4ae822e-eca9-4fcb-8747-217f2a1f53a1",
  "buildId": "0x0f75775f470c1585",
  "components": Object {
    "watch": Object {
      "higgs": Object {
        "filename": "device-higgs.zip",
        "platform": Array [
          "30.1.2+",
        ],
      },
      "meson": Object {
        "filename": "device-meson.zip",
        "platform": Array [
          "32.4.18+",
        ],
      },
    },
  },
  "manifestVersion": 6,
  "requestedPermissions": Array [],
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": Object {
    "device": Object {
      "higgs": "sourceMaps/device/higgs/index.js.json",
      "meson": "sourceMaps/device/meson/index.js.json",
    },
  },
}
`;

exports[`emits an error if a component bundle tag has a device type but invalid platform 1`] = `"Unknown bundle component tag: Invalid value \\"1.1.1+\\" supplied to : (({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>) | { type: \\"companion\\" })/0: ({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>)/0: { type: \\"device\\", family: string, platform: Array<string> }/platform: Array<string>"`;

exports[`emits an error if a component bundle tag has a device type but missing family 1`] = `"Unknown bundle component tag: Invalid value undefined supplied to : (({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>) | { type: \\"companion\\" })/0: ({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>)/0: { type: \\"device\\", family: string, platform: Array<string> }/family: string"`;

exports[`emits an error if a component bundle tag has a device type but missing platform 1`] = `"Unknown bundle component tag: Invalid value undefined supplied to : (({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>) | { type: \\"companion\\" })/0: ({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>)/0: { type: \\"device\\", family: string, platform: Array<string> }/platform: Array<string>"`;

exports[`emits an error if a component bundle tag has an invalid type field 1`] = `"Unknown bundle component tag: Invalid value \\"__invalid__\\" supplied to : (({ type: \\"device\\", family: string, platform: Array<string> } & Partial<{ isNative: true }>) | { type: \\"companion\\" })/type: \\"device\\" | \\"companion\\""`;

exports[`emits an error if both JS and native device components are present 1`] = `"Cannot bundle mixed native/JS device components"`;

exports[`emits an error if multiple bundles are present for the same device family 1`] = `"Duplicate device/foo component bundles: bundle1.zip / bundle0.zip"`;
